qutip/core/dimensions.py

```diff
@@ -27,6 +28,8 @@ def flatten(l):
     [0, 1, 2]
 
     """
+    if isinstance(l, (Space, Dimensions)):
```

Comment:
We should mention in the docstring what the argument `l` can be and what happens with them (Space and Dimension are converted to lists, scalars and other non-lists are converted to single element lists).